ToDo REST API 만들기

jangsc0000 opened the pull request:
Java를 이용해서 ToDo REST API를 만들어봅니다.

ToDo 목록 얻기 - GET /tasks
ToDo 상세 조회하기 - GET /tasks/{id}
ToDo 생성하기 - POST /tasks
ToDo 제목 수정하기 - PUT/PATCH /tasks/{id}
ToDo 삭제하기 - DELETE /tasks/{id}

jangsc0000 committed:
기본적인 HTTP 서버 구현
- Request가 들어오면 method, path, body를 출력
- Response의 body는 항상 "Hi, World!"를 출력

jangsc0000 committed:
Task 클래스 추가
- To Do의 아이탬을 위해 Task 클래스 추가
- Task는 id(long), title(String)을 갖는 클래스

jangsc0000 committed:
엔드포인트 추가 - /tasks GET
- /tasks 엔드포인트의 GET method 추가
- /tasks GET 은 저장된 모든 task를 json포맷으로 반환

jangsc0000 committed:
엔드포인트 추가 - /tasks POST
- /tasks 엔드포인트의 POST method 추가
- /tasks POST는
  1. json으로 입력받은 title을 가지는 Task를 생성
  2. 생성된 Task의 id는 최근에 생성된 Task의 id +1
    2-1. 최근 생성된 Task가 없다면 1

jangsc0000 committed:
엔드포인트 수정 - /tasks POST
- Json parsing 중 에러발생시 staus code 400 반환(빈 입력에 대한 예외처리)
- Task 생성 성공 시 생성된 Task의 json 반환

jangsc0000 committed:
Task method toString 추가

johngrib commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,10 +1,25 @@
 
     public static void main(String[] args) {
+        try {
+            InetSocketAddress address = new InetSocketAddress(8000);
```
`8000`은 어떤 의미를 갖는 값인가요? 이 값을 고정하지 않고 파라미터로 받게 고쳐야 한다면 어떻게 하면 좋을까요?

jangsc0000 replied:
main의 파라미터 args 로 받게 고쳐보라는 말씀이실까요?.. 
어떻게 해야할지 감이 안오네요 ㅠㅠ

johngrib replied:
저는 `8000`을 `final` 변수로 선언하는 것을 생각하고 드린 말씀이긴 했는데요, 정답이 없는 문제이니까 다양하게 시도해보시면 좋겠어요. 방법은 항상 여러가지가 있어요. 여기에서 "정답이 없는"을 꼭 기억해두시고 염두에 두셨으면 좋겠습니다. 어떻게 할지 몰라서 패닉에 빠지는 경우가 된다면 그냥 무작정 코딩을 하기보다는 코드 리뷰를 통해 대화를 시도해보셔도 괜찮아요.

jangsc0000 replied:
넵 감사합니다! config파일로 빼야되나, cli 파라미터로 받아야하나 뭐가 좋을까 계속 고민했는데.. 고민도 리뷰받을수 있겠군요 ㅋㅋㅋ 일단 final로 고쳐볼게요 ㅎㅎ

johngrib replied:
네네 혼자서 고민을 많이 하면 혼란스럽고 때로는 우울해지기도 합니다. 판단이 잘 서지 않을 때에는 꼭 대화를 시도해 보세요.

jangsc0000 replied:
71bde96 final 변수로 고쳐봤습니다

johngrib commented on app/src/main/java/com/codesoom/assignment/MyHttpHandler.java:
```diff
@@ -0,0 +1,87 @@
+
+        String content = "";
+
+        if (method.equals("GET") && path.equals("/tasks")) {
```
만약 `path` 값으로 `/tasks/`가 들어온다면 어떻게 될까요?

jangsc0000 replied:
URI 마지막 문자로 슬래시(/)를 포함하면 안된다고 알고있습니다. - https://meetup.toast.com/posts/92
따라서 마지막 문자로 슬래시(/)를 포함할 경우 /tasks/{id} 엔드포인트에서 처리할 생각입니다.
(404를 날릴지 400을 날릴지는 아직 못정했어요.. ㅠㅠ)

johngrib replied:
URI 설계시에 고려할 문제이긴 한데 너무 깐깐하게 생각하신 것 같아요. 입력으로 들어오는 주소에 대한 것과는 별개의 문제에요. `/tasks`는 받고 `/tasks/`는 받지 않겠다는 것의 의미를 여러가지로 생각해 보세요. 보통은 마지막에 있는 `/`를 떼고 받습니다. 
이건 구현하기 나름이라서, 말씀하신 것과 같이 마지막에 오는 `/`를 잘못된 형식으로 판별하는 것도 선택할 수 있어요.

물론 이런 의사결정은 지금은 이른 편이고, 작업도 번거로우니 지금 코드 상태로도 괜찮아 보여요.

jangsc0000 replied:
감사합니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/MyHttpHandler.java:
```diff
@@ -0,0 +1,87 @@
+            try {
+                Task task = toTask(requestBody);
+
+                long newId = tasks.size() == 0 ? 1 : tasks.get(tasks.size() - 1).getId() + 1;
```
새로운 아이디를 생성하는 코드로군요. 이 코드는 `private` 메소드나 다른 클래스로 책임을 분리해볼 수 있을 것 같네요.

jangsc0000 replied:
17363db TaskManager라는 클래스로 분리하였습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/MyHttpHandler.java:
```diff
@@ -0,0 +1,87 @@
+
+        if (method.equals("GET") && path.equals("/tasks")) {
+            content = tasksToJson();
+            exchange.sendResponseHeaders(200, content.getBytes().length);
```
`200`에 이름이 있으면 좋을 것 같습니다.

jangsc0000 replied:
9c2af92 enum HttpStatusCode를 추가했습니다.

johngrib reviewed (commented):
안녕하세요. 반갑습니다.

jangsc0000 committed:
enum HttpStatusCode 추가

jangsc0000 committed:
엔드포인트 추가 - /tasks/{id} GET
- /tasks/ 엔드포인트도 /tasks/{id} 엔드포인트가 처리하도록 구현(startsWith 사용)
- 상태코드를 enum HttpStatusCode를 사용하도록 모두 변경
- /tasks/{id}가 없거나 id가 숫자가 아닐경우 NOT_FOUND로 구현

johngrib commented on app/src/main/java/com/codesoom/assignment/HttpStatusCode.java:
```diff
@@ -0,0 +1,20 @@
+package com.codesoom.assignment;
+
+public enum HttpStatusCode {
```
이런 enum에는 RFC 2616 문서를 주석으로 남겨주면 어떨까요?

https://datatracker.ietf.org/doc/html/rfc2616#section-10

jangsc0000 replied:
7f47bdf 주석 추가했습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/MyHttpHandler.java:
```diff
@@ -0,0 +1,119 @@
+            }
+        }
+
+        if (method.equals("GET") && path.startsWith("/tasks/")) {
```
각 HTTP 메소드를 처리하는 if 문 안쪽을 각각 private 메소드로 분리해 보세요. 지금 `handle` 메소드의 길이가 너무 길어서 좀 줄여주면 좋을 것 같습니다.

jangsc0000 replied:
17363db 메소드로 분리하였습니다.

jangsc0000 committed:
enum HttpStatus 주석 추가
- RFC2616 section 10의 링크 추가

jangsc0000 committed:
리팩토링
- 각 엔드포인트의 동작들을 createTask, readTask, readTasks 메소드로 분리
- TaskManager를 만들어 List tasks 관리

jangsc0000 committed:
엔드포인트 추가 - /tasks/{id} DELETE

jangsc0000 committed:
엔드포인트 추가 /tasks/{id} PUT/PATCH

johngrib commented on app/src/main/java/com/codesoom/assignment/HttpStatusCode.java:
```diff
@@ -0,0 +1,27 @@
+    Written based on RFC2616 section 10
+    https://datatracker.ietf.org/doc/html/rfc2616#section-10
+*/
+
```
이 한 줄은 삭제해주는 것이 좋습니다.
```suggestion
```

johngrib commented on app/src/main/java/com/codesoom/assignment/HttpStatusCode.java:
```diff
@@ -0,0 +1,27 @@
+package com.codesoom.assignment;
+
+/*
+    Written based on RFC2616 section 10
+    https://datatracker.ietf.org/doc/html/rfc2616#section-10
+*/
```
기왕 주석을 작성한다면 JavaDoc 포맷을 맞추는 것이 좋겠죠. JavaDoc에 대해서도 조사해보시면 좋은 공부가 될 거에요.

```suggestion
/*
 * Written based on RFC2616 section 10
 * https://datatracker.ietf.org/doc/html/rfc2616#section-10
 */
```

johngrib commented on app/src/main/java/com/codesoom/assignment/MyHttpHandler.java:
```diff
@@ -0,0 +1,158 @@
+
+        printRequestInfo(method, path, body);
+
+        if (method.equals("GET") && path.equals("/tasks")) {
```
`GET`, `POST`, `DELETE` 같은 HTTP Method도 enum으로 표현해보면 어떨까요?

jangsc0000 replied:
c9b0f9e HttpMethod enum 추가했습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/MyHttpHandler.java:
```diff
@@ -0,0 +1,158 @@
+        }
+    }
+
+    private void sendResponse(HttpExchange exchange, HttpStatusCode httpStatusCode, String content) throws IOException {
```
`sendResponse`를 호출하는 곳이 꽤 많은 것 같군요. 이 메소드의 역할에 대해 주석으로 설명을 남겨보면 이 코드를 읽는 사람이 더 쉽게 이해할 수 있지 않을까요?

jangsc0000 replied:
2feb54b 주석 추가했습니다.

johngrib reviewed (commented):
한 주간 고생 많으셨습니다. 일요일 저녁에는 주간 회고를 작성해 공유해 주세요.

jangsc0000 commented:
리눅스에서 다시 환경 구성하다가 못하고 뻗어버렸네요..
차주에는 리뷰 주신 거 마저 고치고, 주간 회고도 작성해보겠습니다.
다음주도 잘 부탁드립니다!

jangsc0000 committed:
HttpStatusCode의 javadoc 추가

jangsc0000 committed:
포트의 리터럴을 final 변수로 변경

jangsc0000 committed:
enum HttpMethod 추가
- 모든 메소드의  문자열  리터럴을 enum으로 변경

jangsc0000 committed:
sendResponse 메소드의 주석 추가

jangsc0000 commented:
주간회고 작성했습니다!
https://jangsc0000.github.io/til/%EC%BD%94%EB%93%9C%EC%88%A8-%EC%8A%A4%ED%94%84%EB%A7%8[…]-1%EC%A3%BC%EC%B0%A8-%EC%A3%BC%EA%B0%84%ED%9A%8C%EA%B3%A0/
